### DISCUSSION TITLE
What is the correct way of adding a menu without having separate .ts files?

### DISCUSSION THREAD

**Discussion by lorenzwoehr** [yuanqing/create-figma-plugin#165] [Q&A] [answered]
My plan is to add several commands to the plugin menu, which call different functions in `main.ts`.
However, currently the `"main"` field in `figma-plugin` gets converted to `"command"`. 

For instance ... 

```
"menu": [
      {
        "name": "Command 1",
        "main": "src/main.ts",
        "ui": "src/ui.tsx"
      },
      "-",
      {
        "name": "Command 2",
        "main": "src/main.ts",
        "ui": "src/ui.tsx"
      }
]
```
    
... gets converted to the following in `manifest.json` : 
```
"menu": [
    {
      "name": "Command 1",
      "command": "src/main.ts--default"
    },
    {
      "separator": true
    },
    {
      "name": "Command 2",
      "command": "src/main.ts--default"
    }
]
```

When I'm trying to add the following directly in `manifest.json` , Figma complains about **Uncaught Error: No UI defined for command `command-1`**, presumably because I'm not calling `figma.showUI()` in `main.ts `. Instead I'm calling `showUI()` from the `create-figma-plugin` package:

```
"main": "build/main.js",
  "ui": "build/ui.js",
  "menu": [
    {
      "name": "Command 1",
      "command": "command-1"
    },
    {
      "separator": true
    },
    {
      "name": "Command 2",
      "command": "command-2"
    }
]
```

Thank you in advance 🙏

**Comment by yuanqing** [comment, marked as answer]
> add several commands to the plugin menu, which call different functions in `main.ts`

Instead of only using the `default` export for all the commands, you can specify a particular named export for each command instead. For example:

```json
"menu": [
  {
    "name": "Command 1",
    "main": {
      "src": "src/main.ts",
      "handler": "commandOne"
    },
    "ui": "src/ui.tsx"
  },
  "-",
  {
    "name": "Command 2",
    "main": {
      "src": "src/main.ts",
      "handler": "commandTwo"
    },
    "ui": "src/ui.tsx"
  }
]
```

Your `src/main.ts` can then be something like:

```ts
export function commandOne () {
  // ...
}

export function commandTwo () {
  // ...
}
```

> When I'm trying to add the following directly in `manifest.json`

I do not recommend manually editing the `manifest.json` because it will always get overridden whenever you rebuild your plugin
